### PR TITLE
Custom regex based solution.

### PR DESCRIPTION
**Pull request by Miro-Andrin**
The way command dispatching is done is by using regex to quickly determine what plugin has a command that probably is able to parse/handle the given input. The input is then written into the plugin's memory were it then tries to parse it.

The custom regex stuff we are doing lets us have early termination. What that means is that when we have two commands.
/tp ...and /tpa , then the dispatcher would know what command to dispatch after at most seeing 3 chars. /tp[^a] -> first command,/tpa -> second command. This early termination is figured out at runtime, so it works for plugins loaded later on. Using our custom regex stuff we can add some functionality we cant get whit the default regex crate. We can determine what commands probably try to handle the same input. That way we can detect cases were let's say two plugins want a teleport command with the syntax/tp ... . With some tinkering we probably can get tab completion working based on the regex, such that whenever someone presses tab and the regex says only one string of chars is valid, then its auto completes based on that.

 We also have a command/parser builder pattern. It builds a parser/command, but also keeps track of the regex that is associated with the given parser. Therefore anyone not implementing their own argument type (like a custom target selector syntax), wont have to know about the existence of the regex. 
 
 Defman and I started on a custom regex engine some months ago. Recently i returned to it, rewrote most of the existing code, and added a few necessary features. So anything that does not seem idiomatic is probably my doing.

### CODE REVIEW

File: src/parser/literal.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                            "Next char:'{}' was not whitespace or 'end of str' after literal.",
+                            c
+                        )),
+                        None,
+                    );
+                }
+                (Some(_), None) => {
+                    return (
+                        Err(anyhow!(
+                            "The input: \"{}\" is shorter then the literal: \"{}\"",
```
Comment (caelunshun) [inline src/parser/literal.rs:56]: than

File: src/parser/literal.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        _state: Self::ParserState,
+        input: &'i str,
+    ) -> (Result<(Self::Extract, &'i str)>, Option<Self::ParserState>) {
+        let input_lower = &mut input
+            .trim_start()
+            .chars()
+            .flat_map(|c| c.to_lowercase())
+            .peekable();
+        let literal_lower = &mut self.value.chars().flat_map(|c| c.to_lowercase()).peekable();
+        let mut ofsett = 0;
```
Comment (caelunshun) [inline src/parser/literal.rs:31]: Should this be `offset`?

**Review by caelunshun — COMMENTED** [review]
Thanks! On my first review, this looks good—just a few typos. Some of the regex code goes over my head, but the tests look thorough enough to me.

My main concern here is that this is much more complex than the current implementation or even Brigadier itself, and I'm not sure all that complexity is really necessary. What's the motivation for the custom regex engine?

Looking at Brigadier's code, it uses relatively simple graph traversals. As Lieutenant is supposed to reimplement Brigadier, I'm of the opinion we should have a similar implementation. If the regex solution provides substantial benefits over the simpler algorithm, then I'm all for it; it's jut not yet clear to me what those advantages are.

**Comment by Miro-Andrin** [comment]
> What's the motivation for the custom regex engine?

I would be lying if I did not mention that it's somewhat due to it being fun to code 😃, but there were some less egotistical reasons as well.

a) Looking at Argument<C>'s PartialEq implementation we can see its relies on argument_type: TypeId to distinguish between Argument::Parser's. This is as far as I understand not sound when we are dealing with arguments that can come from different Plugins using different rust compiler versions.
So we at least need some other way to determine if parsers are identical.
I however think we need to go a step further and have a way of determining if two parsers have overlap in what they recognise (not by where they were created). This would give better diagnostics.  Which is good for users, and  plugin developers.
I think Mojang have realized this issue as well, because when you use their data generators they give warnings about ambiguity in their commands.

```
[14:34:13] [main/WARN]: Ambiguity between arguments [teleport, destination] and [teleport, targets] with inputs: [Player, 0123, @e, dd12be42-52a9-4a91-a8a1-11c01849e498]
[14:34:13] [main/WARN]: Ambiguity between arguments [teleport, location] and [teleport, destination] with inputs: [0.1 -0.5 .9, 0 0 0]
[14:34:13] [main/WARN]: Ambiguity between arguments [teleport, location] and [teleport, targets] with inputs: [0.1 -0.5 .9, 0 0 0]
[14:34:13] [main/WARN]: Ambiguity between arguments [teleport, targets] and [teleport, destination] with inputs: [Player, 0123, dd12be42-52a9-4a91-a8a1-11c01849e498]
[14:34:13] [main/WARN]: Ambiguity between arguments [teleport, targets, location] and [teleport, targets, destination] with inputs: [0.1 -0.5 .9, 0 0 0]
```
Using the custom regex solution we get this partial equality check.

b) I believe custom regex is the best way of handling tab completion. The internal data structures (DFA) of the regex engine gives this almost for free.  Maybe you want to see a working example, and not just future promises?

c) The regex solution is also probably faster, although not measured, and also not as important due it the existing solution being good enough.

I agree on you comment on complexity. However, I think some complexity is necessary. The way my code is now it can be though of as two separate crates  (that actually probably should be split up into two distinct crates). We have the regex stuff in '/src/regex' and the parser combinator stuff in the other folders/modules.

There is a lot of code connected to  handling regex. But if you have the mental model of how a DFA and NFA works, its actually straight forward and relatively simple code. It is more abstract, but conceptually simple once you know the model. If you were convinced by points a-b, then you can understand why I think this type of complexity and abstraction is necessary and god. 

The complexity that needs revisiting however is everything else. Especially the parsers. 
```rust
pub trait IterParser {
    
    type Extract: Tuple;
    type ParserState: Default;

    #[allow(clippy::type_complexity)]
    fn parse<'p>(
        &self,
        state: Self::ParserState,
        input: &'p str,
    ) -> (
        anyhow::Result<(Self::Extract, &'p str)>,
        Option<Self::ParserState>,
    );

    fn regex(&self) -> String;
}
```
The reason the parse trait got so complicated was that I wanted to sport optionals. That meant that we needed to be able to backtrack. This made it necessary that parsers be generators, that then required the ParserState. I could however say we don't sport optionals that way, and just require users to submit two commands. One for the None case and one for the Some case.  This would greatly simplify the parser code. 

```rust
pub trait IterParser {
    
    type Extract: Tuple;

    fn parse<'p>(
        &self,
        input: &'p str,
    ) -> anyhow::Result<(Self::Extract, &'p str)>;

    fn regex(&self) -> String;
}
```

**Comment by caelunshun** [comment]
Sorry about the late response—but I understand why the regex solution is beneficial now. I still think it's _slightly_ overengineered, but since all the code's already implemented, I see no reason to hold off on the complexity.

On the tab completion point: with the regex-based implementation, we get some functionality for free, but what if plugins need custom tab completion? For example, if an argument should be an online player's username? Maybe I'm not understanding this fully, but it seems the regex won't allow for this sort of higher-level completion, or at least it'll be nontrivial to implement.

In any case, it'll be good to have at least basic command support in Feather, so I'll go ahead and merge this. Thank you and @Defman for all the hard work here!

**Review by caelunshun — APPROVED** [review]

**Pull request merged**